src/lseq/ident.rs

```diff
@@ -0,0 +1,409 @@
+            path: vec![(2u64.pow(self.initial_base_bits) - 1, None)],
+        }
+    }
+
+    /// Allocates a new identifier between `p` and `q` such that `p < z < q`.
+    ///
+    /// The way to think about this is to consider the problem of finding a short number between
+    /// two decimal numbers. For example, between `0.2` and `0.4` we'd like to return `0.3`.
+    /// But what about `0.2` and `0.3`? Well a nice property is that if we add a digit to `0.2`, the
+    /// number produced will be larger than `0.2` but smaller than `0.3`.
+    ///
+    /// If you view identifiers `p` and `q` as decimal numbers `0.p` and `0.q` then all we're doing is
+    /// finding a number between them!
+    ///
+    /// # Panics
+    ///
+    /// * `p` equal to `q`.
+    /// * If no identifier of length less than 31 exists between `p` and `q`.
+    pub fn alloc(&mut self, p: &Identifier<A>, q: &Identifier<A>) -> Identifier<A> {
+        assert!(p != q, "allocation range should be non-empty");
+        if q < p {
+            self.alloc(q, p);
+        }
+
+        let mut depth = 0;
+
+        // Descend both identifiers in parallel until we find room to insert an identifier
+        loop {
+            match (p.path.get(depth), q.path.get(depth)) {
+                // Our descent continues...
+                (Some(a), Some(b)) => {
+                    if a.0 == b.0 {
+                        // continue searching lower
+                    } else if a.0 + 1 < b.0 {
+                        // We take this branch if there's room between a and b to insert a new
+                        // identifier
+                        let next_index = self.index_in_range(a.0 + 1, b.0, depth as u32);
+                        return self.replace_last(p, depth, next_index);
+                    } else {
+                        // If we end up here that means that a + 1 = b. Therefore we can't actually
+                        // insert a new node at this layer. So we start a new layer below a.
+                        return self.alloc_from_lower(p, depth + 1);
+                    }
+                }
+
+                // The upper bound ended on the previous level.
+                // This means we can allocate a node in the range a..max_for_depth
+                // If there isn't room at this level because a = max_for_depth then we'll keep
+                // searching below.
+                (Some(_), None) => {
+                    return self.alloc_from_lower(p, depth);
+                }
+                // The lower bound ended on the previous level, we just need to keep following the upper bound until some space
+                // is available.
+                (None, Some(_)) => {
+                    return self.alloc_from_upper(p, q, depth);
+                }
+
+                // The two paths are fully equal which means that the site_ids MUST be different or
+                // we are in an invalid situation
+                (None, None) => {
```

Comment:
Isn't this protected against in https://github.com/rust-crdt/rust-crdt/pull/74/files#diff-059cbb2a21b4032a5d9183f7077f5129R95?

Reply:
Nvm, I get it now, the entry at the previous depth should have been from different sites, we should assert that